libgnucash/engine/ScrubBudget.c

```diff
@@ -0,0 +1,211 @@
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+
+#include <config.h>
+
+#include <glib.h>
+#include <glib/gi18n.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "gnc-prefs.h"
+#include "gnc-budget.h"
+#include "gnc-features.h"
+#include "ScrubBudget.h"
+
+static QofLogModule log_module = "gnc.engine.scrub";
+
+typedef enum
+{
+    HEURISTICS_INC_EXP,
+    HEURISTICS_CREDIT_ACC,
+    HEURISTICS_NONE
+} SignReversals;
+
+typedef struct
+{
+    GncBudget* budget;
+    SignReversals policy;
+} ReversalType;
+
+typedef struct
+{
+    gint asset,liability,equity,income,expense;
+    gint num_periods;
+    GncBudget* budget;
+} ProcessData;
+
+static void
+process_heuristics_acct (Account * account, gpointer user_data)
+{
+    /* each account- check type. sum budget period amounts. if sum<0,
+       decrease type tally by 1. if sum>0, increase type tally by 1. */
+    ProcessData *heuristics = (ProcessData*) user_data;
+    gnc_numeric total = gnc_numeric_zero(), val;
+    gint sign;
+    gchar *totalstr;
+
+    for (gint i = 0; i < heuristics->num_periods; ++i)
+    {
+        if (!gnc_budget_is_account_period_value_set (heuristics->budget, account, i))
+            continue;
+        val = gnc_budget_get_account_period_value (heuristics->budget, account, i);
+        total = gnc_numeric_add_fixed (total, val);
+    }
+
+    sign = gnc_numeric_compare (total, gnc_numeric_zero ());
+    totalstr = gnc_numeric_to_string (total);
+    PINFO ("acct=%s, total=%s, sign=%d",
+           xaccAccountGetName (account), totalstr, sign);
+    g_free (totalstr);
+
+    switch (xaccAccountTypeGetFundamental (xaccAccountGetType (account)))
+    {
+    case ACCT_TYPE_ASSET:
+        heuristics->asset += sign;
+        break;
+    case ACCT_TYPE_LIABILITY:
+        heuristics->liability += sign;
+        break;
+    case ACCT_TYPE_EXPENSE:
+        heuristics->expense += sign;
+        break;
+    case ACCT_TYPE_INCOME:
+        heuristics->income += sign;
+        break;
+    case ACCT_TYPE_EQUITY:
+        heuristics->equity += sign;
+        break;
+    default:
+        break;
+    }
+}
+
+static SignReversals
+heuristics_on_budget (GncBudget * budget, Account *root)
+{
+    ProcessData heuristics = {0, 0, 0, 0, 0, gnc_budget_get_num_periods (budget),
+                              budget};
+    SignReversals result;
+
+    gnc_account_foreach_descendant (root, process_heuristics_acct, &heuristics);
+
+    result =
+        heuristics.expense < 0 ? HEURISTICS_INC_EXP :
+        heuristics.liability > 0 ? HEURISTICS_NONE :
```

Comment:
I suggest we use income rather than liability in the second step of heuristics. This is because a liability check fails when budgeting for a new loan such as mortgage.

In summary:
First we check for expense. If it is negative then we are certain that the setting is reverse income / expense.
Now we just need to distinguish "reverse credit account" and "none", where income is one of them. A positive income means "reverse credit account" is set so we should reverse it.

Reply:
I think you're right fwiw

Reply:
On second thoughts, I do not think I will be able to continue this line of work; I do not use budgets at all, and will not want to handle users' bug reports after the budgets fix. I'll be keen to revert this budget scrubbing altogether, including the 4.x handling of unreversed budgets. I'll be happy to reinstate a PR to document the work done, and perhaps someone else may take up the work.

Reply:
On third thoughts, I think I found the bug in this branch, see f7ee6446 and 85a62c03 just merged in. Please test master again.